hudi-common/src/main/java/org/apache/hudi/common/engine/HoodieMREngineContext.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.common.engine;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hudi.common.config.SerializableConfiguration;
+import org.apache.hudi.common.function.SerializableConsumer;
+import org.apache.hudi.common.function.SerializableFunction;
+import org.apache.hudi.common.function.SerializablePairFunction;
+import org.apache.hudi.common.util.Option;
+
+import org.apache.hudi.common.util.collection.Pair;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static java.util.stream.Collectors.toList;
+import static org.apache.hudi.common.function.FunctionWrapper.throwingFlatMapWrapper;
+import static org.apache.hudi.common.function.FunctionWrapper.throwingForeachWrapper;
+import static org.apache.hudi.common.function.FunctionWrapper.throwingMapToPairWrapper;
+import static org.apache.hudi.common.function.FunctionWrapper.throwingMapWrapper;
+
+/**
+ * A java based engine context that can be used from map-reduce tasks executing in query engines like
+ * spark, hive and presto.
+ */
+public final class HoodieMREngineContext extends HoodieEngineContext {
```

Comment:
any reason why this cannot be in `hudi-hadoop-mr`?

Reply:
Well it is being directly instantiated at a couple of places in `hudi-common`:
1. `FileSystemViewManager`: https://github.com/apache/hudi/pull/2417/files#diff-1fa21ae97e976db303c47571e7cb9c34e4f94719d482e537134e231b2e944388R167
2. `HoodieBackedTableMetadata`: https://github.com/apache/hudi/pull/2417/files#diff-7c43aea81a02b4f135452b50eaa36d5868081e72b37d43101ca9de1f9ebb5195R77

`hudi-common` does not depend on `hudi-hadoop-mr`.  As per your other comment we can get rid of `1` if we pass the engine context from the caller side. But we should not get rid of `2` because we are treating `HoodieEngineContext` as the default engine context to instantiate. This constructor is being used from `hud-cli` for ex (does not depend on `hudi-hadoop-mr`:
-  https://github.com/apache/hudi/pull/2417/files#diff-b9829aa53bc2a7b0a655993df2c3c75d53f6a8e51f317753e51f27c63fc20341R165
- https://github.com/apache/hudi/pull/2417/files#diff-b9829aa53bc2a7b0a655993df2c3c75d53f6a8e51f317753e51f27c63fc20341R197

So, if we want to treat it like a default engine context to be used would be better to keep in `hudi-common`.

Reply:
> But we should not get rid of 2 because we are treating HoodieEngineContext as the default engine context to instantiate

But that only depends on the interface right. `hudi-cli` has access to sparkContext right?

Reply:
Yes but not all commands need spark context. It is created depending on the commands which truly need spark context.

Besides this, as discussed offline there is another issue which I encountered when I started to fetch `hadoopConf` from `engineContext` based on your other comment. It results in `NPE` at places when metadata table is created inside spark executor tasks because `engineContext` is not serializable and ends up null at executors. For such scenarios in `HoodieTable` where we create metadata table I will have to add a check to create a `HoodieMREngineContext` in case engine context is `null`. Thus `hudi-client` needs access to this as well. So based on what we agreed to, we need something like a `HoodieLocalEngineContext` that can be used at all these places and can sit in `hudi-common`.